### PR TITLE
applications: Added ESL AP and Tag application

### PR DESCRIPTION
**Pull request by pirun**
ESL tag sample code.
LED tag could use all of Nordic Devkit.
EPD tag could use nRF52832, nRF52833.

### CODE REVIEW

**Comment by jori-nordic** [comment]
@pirun could you move everything into `sdk-nrf/applications`? Including the services.

**Comment by pirun** [comment]
> @pirun could you move everything into `sdk-nrf/applications`? Including the services.

For sure. I'll submit a merged revision.

File: applications/esl/doc/esl_ap.rst [hunk 1]
```diff
@@ -0,0 +1,15 @@
+.. _esl_ap:
+
+ESL Access Point
+################
+
+The ESL Access Point sample demonstrates how to use the :ref:`esl_service_client_readme`.
```
Comment (jori-nordic) [inline applications/esl/doc/esl_ap.rst:6]: `sample` -> `application`
for all the occurrences

File: applications/esl/peripheral_esl/hw/display/epd_shield_display.c [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+#if defined(ESL_POWER_PROFILE)
+	display_gpio_onoff(true);
+	/**
+	 * To optimize power remove static declaration of ssd16xx_init in
+	 * zephyr/drivers/display/ssd16xx.c
+	 **/
+	ssd16xx_init(display_dev);
+	(void)pm_device_action_run(spi, PM_DEVICE_ACTION_RESUME);
+#endif
+#if (CONFIG_BT_ESL_PAINT_LIB)
```
Comment (jori-nordic) [inline applications/esl/peripheral_esl/hw/display/epd_shield_display.c:225]: where is this lib from? I don't see it in zephyr or sdk-nrf

File: doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst [hunk 3]
```diff
@@ -255,6 +260,10 @@ Bluetooth samples
   * Changed:
 
     * Fixed an interoperability issue with iOS devices by setting the report IDs of HID input and output reports to zero.
+* Added:
+
+  * :ref:`central_esl` sample.
```
Comment (jori-nordic) [inline doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst:265]: sample -> application

File: doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst [hunk 4]
```diff
@@ -90,6 +90,11 @@ Bluetooth® LE
   The previous ``hci_err.h`` header has been merged into the new ``hci_types.h`` header.
   This can break builds that were directly including ``hci_err.h``.
 
+* Added:
+
+  * :ref:`esl_service_readme` library.
+  * :ref:`esl_service_client_readme` library.
```
Comment (jori-nordic) [inline doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst:96]: You should remove those since they're not usable outside of the esl application now

File: applications/esl/doc/esl_ap.rst (hunk 1 above)
Comment (pirun) [reply applications/esl/doc/esl_ap.rst:6]: Change for all the occurrences.

File: doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst (hunk 3 above)
Comment (pirun) [reply doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst:265]: for all the occurrences

File: doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst (hunk 4 above)
Comment (pirun) [reply doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst:96]: Remove

File: applications/esl/peripheral_esl/hw/display/epd_shield_display.c (hunk 2 above)
Comment (pirun) [reply applications/esl/peripheral_esl/hw/display/epd_shield_display.c:225]: Discussion offline.

File: applications/esl/peripheral_esl/hw/esl_hw_impl.c [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+
+#include <zephyr/logging/log.h>
+
+#include "esl.h"
+
+LOG_MODULE_DECLARE(peripheral_esl);
+
+void hw_chrc_init(struct bt_esl_init_param *init_param)
+{
+	/** Initial Display elements here **/
```
Comment (alwa-nordic) [inline applications/esl/peripheral_esl/hw/esl_hw_impl.c:15]: typo

File: applications/esl/peripheral_esl/hw/esl_hw_impl.c [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+	for (uint8_t idx = 0; idx < CONFIG_BT_ESL_DISPLAY_MAX; idx++) {
+		init_param->displays[idx].width = CONFIG_ESL_DISPLAY_WIDTH;
+		init_param->displays[idx].height = CONFIG_ESL_DISPLAY_HEIGHT;
+		init_param->displays[idx].type = CONFIG_ESL_DISPLAY_TYPE;
```
Comment (alwa-nordic) [inline applications/esl/peripheral_esl/hw/esl_hw_impl.c:19]: wouldn't it make more sense to have CONFIG_ESL_DISPLAY_0_WIDTH etc? Or use device tree for this?

File: applications/esl/peripheral_esl/hw/esl_hw_impl.h [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+
+#ifndef EL_HW_IMPL_H_
+#define EL_HW_IMPL_H_
+#include <zephyr/device.h>
+
+/**
+ * @brief Header for format generated by Image2LCD for Waveshare EPD
+ * Option include head data.
+ * Scan mode Data hor,Byte ver to match Zephyr ssd16xx driver SSD16XX_DATA_ENTRY_XDYIY
+ */
```
Comment (alwa-nordic) [inline applications/esl/peripheral_esl/hw/esl_hw_impl.h:15]: Document the endianess of w and h here. And the meaning of scan and gray.

File: applications/esl/peripheral_esl/hw/esl_hw_impl.h [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+int ots_storage_init(void);
+
+/**
+ * To optimize power remove static declaration of ssd16xx_init in zephyr/drivers/display/ssd16xx.c
+ **/
+int ssd16xx_init(const struct device *dev);
+/**
+ * To optimize power remove static declaration of uc81xx_init in zephyr/drivers/display/uc81xx.c
+ **/
+int uc81xx_init(const struct device *dev);
```
Comment (alwa-nordic) [inline applications/esl/peripheral_esl/hw/esl_hw_impl.h:46]: Is there a header that can be included instead of declaring here?

File: applications/esl/peripheral_esl/hw/led/led.c [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+		LOG_ERR("Device %s not found", LED_PWM_DEV_NAME);
+		return -ENOENT;
+	}
+
+	return 0;
+}
+void led_control(uint8_t led_idx, uint8_t color_brightness, bool onoff)
+{
+	int err;
+	uint8_t level = (((color_brightness & 0xC0) >> ESL_LED_BRIGHTNESS_BIT) + 1) * 25;
```
Comment (alwa-nordic) [inline applications/esl/peripheral_esl/hw/led/led.c:43]: This needs an explanatory comment.

File: applications/esl/peripheral_esl/hw/sensor/sensor.c [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+
+	int32_t celsius_temperature = nrfx_temp_calculate(temperature);
+	int32_t whole_celsius = celsius_temperature / 100;
+	uint8_t fraction_celsius = NRFX_ABS(celsius_temperature % 100);
+
+	atomic_set_bit(&esl_obj->basic_state, ESL_SERVICE_NEEDED_BIT);
+	esl_obj->sensor_data[0].data_available = true;
+	LOG_INF("Measured temperature: %d.%02u [C], raw %d", whole_celsius, fraction_celsius,
+		celsius_temperature);
+	sys_put_le16((uint16_t)celsius_temperature, &esl_obj->sensor_data[0].data[0]);
```
Comment (alwa-nordic) [inline applications/esl/peripheral_esl/hw/sensor/sensor.c:33]: Please don't use `sys_put_le16` for non-uint16_t.

File: applications/esl/peripheral_esl/conf/nrf52840dk_nrf52840/nrf52840dk_nrf52840_epd.overlay [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+// 			partition = <&lfs1_part>;
+// 			automount;
+// 			read-size = <16>;
+// 			prog-size = <16>;
+// 			cache-size = <64>;
+// 			lookahead-size = <32>;
+// 			block-cycles = <512>;
+// 		};
+// 	};
+// };
```
Comment (alwa-nordic) [inline applications/esl/peripheral_esl/conf/nrf52840dk_nrf52840/nrf52840dk_nrf52840_epd.overlay:38]: dead

File: applications/esl/peripheral_esl/hw/esl_hw_impl.h [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+ */
+
+#ifndef EL_HW_IMPL_H_
+#define EL_HW_IMPL_H_
+#include <zephyr/device.h>
+
+/**
+ * @brief Header for format generated by Image2LCD for Waveshare EPD
+ * Option include head data.
+ * Scan mode Data hor,Byte ver to match Zephyr ssd16xx driver SSD16XX_DATA_ENTRY_XDYIY
```
Comment (alwa-nordic) [inline applications/esl/peripheral_esl/hw/esl_hw_impl.h:14]: typo

**Review by alwa-nordic — APPROVED** [review]

File: applications/esl/README.rst [hunk 13]
```diff
@@ -0,0 +1,23 @@
+.. _esl_application:
+
+Electronic Shelf Label
+######################
+
+.. contents::
+   :local:
+   :depth: 2
+
+The Electronic Shelf Label application provides both Access Point and Tag role.
```
Comment (FrancescoSer) [inline applications/esl/README.rst:10]: Could you add some more context to what kind of objective is this application trying to achieve?

Comment (pirun) [reply applications/esl/README.rst:10]: Add more content, please review it.

**Review by greg-fer — CHANGES_REQUESTED** [review]
Has the documentation for this application been discussed with a tech writer before the PR was opened? Because it ignores the application template we have https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/templates/application_README.html and it does not follow the structure of the existing applications.

Looks like the case similar to the Gazell samples that were bundled together and presented as one sample, but in this case it's an application.

Also looking at the existing doc files, there is no mention of DFU, which is one of the product requirements for an application. There is also no information about how to integrate your own hardware with the application, typical for several applications we have.

In short: this requires a major rework of the documentation and possibly of the code part. I would suggest this PR be taken to draft and reworked before any documentation is reviewed.

**Comment by pirun** [comment]
Plan to go to nordicplayground

**Comment by pirun** [comment]
Open temporary for ducoment sharing.
